src/operator/subgraph/mkldnn/mkldnn_fc_sum_fuse.h

```diff
@@ -0,0 +1,240 @@
+ public:
+  SgMKLDNNFCSumFuseProperty() {}
+
+  static SubgraphPropertyPtr Create() {
+    static const std::string &name = "MKLDNN FullyConnected post quantization second pass";
+    auto property = std::make_shared<SgMKLDNNFCSumFuseProperty>();
+    property->SetAttr<std::string>("property_name", name);
+    property->SetAttr<bool>("inference_only", true);
+    if (dmlc::GetEnv("MXNET_DISABLE_MKLDNN_FC_SUM", 0)) {
+      property->SetAttr<bool>("disable", true);
+    }
+    return property;
+  }
+
+  nnvm::ObjectPtr CreateSubgraphNode(const nnvm::Symbol &sym,
+                                   const int subgraph_id = 0) const override {
+    nnvm::ObjectPtr fc_node = nullptr;
+    nnvm::ObjectPtr ew_add_node = nullptr;
+
+    DFSVisit(sym.outputs, [&](const nnvm::ObjectPtr &node) {
+      if (node->is_variable()) return;
+      auto &sub_name = node->op()->name;
+      if (sub_name == "_sg_mkldnn_fully_connected") {
+        fc_node = node;
+      } else if (sub_name == "elemwise_add") {
+        ew_add_node = node;
+      }
+    });
+
+    CHECK_NOTNULL(fc_node);
+    if (ew_add_node != nullptr) {
+      CHECK_NOTNULL(fc_node->attrs.subgraphs[0]);
+      auto fc_orginal = fc_node->attrs.subgraphs[0]->outputs[0].node;
+      if (fc_orginal->op() == Op::Get("FullyConnected")) {
+        nnvm::Symbol new_sym;
+        nnvm::NodeEntry &ew_input_with_fc = (ew_add_node->inputs[1].node == fc_node) ?
+                                        ew_add_node->inputs[1] :
+                                        ew_add_node->inputs[0];
+        ew_input_with_fc.node = fc_orginal;
+        new_sym.outputs.emplace_back(ew_add_node);
+        fc_node->attrs.subgraphs.clear();
+        fc_node->attrs.subgraphs.emplace_back(std::make_shared<nnvm::Symbol>(new_sym));
+        fc_node->attrs.dict["with_sum"] = "True";
+        fc_node->op()->attr_parser(&(fc_node->attrs));
+      }
+    }
+    return fc_node;
+  }
+
+  SubgraphSelectorPtr CreateSubgraphSelector() const override {
+    bool quantized = HasAttr("quantize") ? GetAttr<bool>("quantize") : false;
+    auto selector =
+      std::make_shared<SgMKLDNNFCSumFuseSelector>(quantized);
+    return selector;
+  }
+
+  void ConnectSubgraphOutputs(
+      const nnvm::ObjectPtr n,
+      std::vector<nnvm::NodeEntry *> *output_entries) const override {
+    // Connect all extern output entries to output[0]
+    for (size_t i = 0; i < output_entries->size(); ++i) {
+      auto entry_ptr = output_entries->at(i);
+      *entry_ptr = nnvm::NodeEntry{n, entry_ptr->index, 0};
+    }
+  }
+
+  void ConnectSubgraphInputs(
+      const nnvm::ObjectPtr n, std::vector<nnvm::NodeEntry *> *input_entries,
+      std::vector<nnvm::NodeEntry> *orig_input_entries) const override {
+    auto sym = n->attrs.subgraphs[0];
+    auto const &fc_param = nnvm::get<MKLDNNFCFullParam>(n->attrs.parsed);
+    std::unordered_set<const nnvm::Node *> node_sets;
+    DFSVisit(sym->outputs, [&](const nnvm::ObjectPtr &node) {
```

Comment:
I think so, but I think it is to big change in this moment.

Reply:
@anko-intel Thanks, How about talking it over, here or as RFC? 

## Problem statement ##
If the structure of a graph is still the same all the time (apart of fusing). Then we can try to avoid running DFS algorithm multiple times.  For brevity, those calls force us to pay a penalty: `O(V + E)`. Assuming that `n` represents a total number of patterns, then the summary cost of that is equal `n * O(V +E)`. Even more, for a few operators (i.e) DFS might be run twice or more, if any ~ then we got `mn * O(V +E)`. 

## Solution - short description ##
Reduce overall time of using DFS to a minimum: nice to have: O(V +E), by labelling nodes and keeping traversal order. Pattern matching is applied after that ~ on a traversal order vector (maybe LCA knowledge is needed).

Reply:
@anko-intel Thanks, How about talking it over? 

## Problem statement ##
If the structure of a graph is still the same all the time (apart of fusing). Then we can try to avoid running DFS algorithm multiple times.  For brevity, those calls force us to pay a penalty: O(V + E). If we assume that `n` represents a total number of patterns, then the summary cost of that is equal n * O(V +E). Even more, for a few operators (i.e) DFS might be run twice or more, if any ~ then we got mn * O(V +E). 

## Solution - short description ##
Reduce overall time of using DFS to a minimum: nice to have just O(V +E), by labelling nodes and keeping traversal order. Pattern matching is applied after that ~ on a traversal order vector (maybe LCA knowledge is needed).